File: src/Components/Deadline.cs
using Ssg.IO;

using Ssg.Utils;

namespace Ssg.Components;

public class Deadline : IComponent
{
    private readonly string id;

    public Deadline(string id) => this.id = id;

    public void OutputRequirements(IWriter writer) =>
        writer.Write("<link rel='stylesheet' type='text/css' href='/req/components/deadline.css'>");

    public void Output(IWriter writer)
    {
        var next = PostsXmlFinder.GetInstance().GetNextOf(this.id);
        var prev = PostsXmlFinder.GetInstance().GetPrevOf(this.id);

        new Tombstone().Output(writer);
        new Node("hr").Output(writer);
        new Node("div")
            .AddAttribute("class", "deadline-links")
            .AddChild(next != null ? this.createLink(next, "Next Article") : new Node("div"))
            .AddChild(prev != null ? this.createLink(prev, "Prev Article") : new Node("div"))
            .Output(writer);
    }

    private IComponent createLink(PostData postData, string text) =>
        new Node("div")
            .AddChild(new Node("div").SetInnerText(text))
            .AddChild(new Node("div").AddChild(new Node("a").AddAttribute("href", $"/posts/{postData.Id}/").SetInnerText(postData.Title)));
}

File: src/coms/deadline.js
const parser = require("../parser")
const posts = require("../posts")

const CSS = `<link rel="stylesheet" type="text/css" href="/coms/deadline.css?20240212">`

exports.run = function(text, eohs) {
  const matcheds = parser.parseSelfClosingTag(text, "Deadline")
  if (matcheds.length === 0) {
    return text
  }

  for (const matched of matcheds) {
    const key = parser.getAttributeOr(matched.attributes, "key", "")
    const next = posts.getNextDataFromKey(key)
    const prev = posts.getPrevDataFromKey(key)
    const nextDiv = next ? `<div><div>Next Article</div><div><a href="/posts/${next.key}">${next.title}</a></div></div>` : "<div></div>"
    const prevDiv = prev ? `<div><div>Prev Article</div><div><a href="/posts/${prev.key}">${prev.title}</a></div></div>` : "<div></div>"
    text = text.replace(matched.entire, `<div class="deadline">${nextDiv}${prevDiv}</div>`)
  }
  
  eohs.push(CSS)
  return text
}
